File: src/core/representation/engine.ts
/**
 * Representation Engine
 * Core abstraction for content representation
 */

import type {
  Representation,
  SimilarityStrategy,
  CrossModelResult,
  EngineCapabilities
} from './types';

/**
 * Core abstraction for content representation.
 * Implementations handle encoding, comparison, serialization.
 */
export interface RepresentationEngine {
  /** Model identifier */
  readonly modelId: string;

  /** Engine capabilities */
  readonly capabilities: EngineCapabilities;

  /**
   * Encode content into representation.
   */
  encode(content: string): Promise<Representation>;

  /**
   * Batch encode for efficiency.
   */
  encodeBatch(contents: string[]): Promise<Representation[]>;

  /**
   * Compute similarity between representations.
   */
  similarity(
    a: Representation,
    b: Representation,
    strategy?: SimilarityStrategy
  ): number;

  /**
   * Compare representations from different models.
   */
  crossModelSimilarity(
    a: Representation,
    b: Representation
  ): CrossModelResult;

  /**
   * Serialize for persistence.
   */
  serialize(rep: Representation): Uint8Array;

  /**
   * Deserialize from persistence.
   */
  deserialize(bytes: Uint8Array): Representation;

  /**
   * Check if representation is from this engine's model.
   */
  isNative(rep: Representation): boolean;
}

/**
 * Engine configuration.
 */
export interface EngineConfig {
  type: 'transformer' | 'mock';
  modelId?: string;
  dimensions?: number;
  /** Quantization for transformer engine */
  dtype?: 'fp32' | 'fp16' | 'q8' | 'q4';
}

export const DEFAULT_ENGINE_CONFIG: EngineConfig = {
  type: 'mock',
  modelId: 'mock-engine-v1',
  dimensions: 384
};

File: crates/evolve-core/src/representation/engine.rs
use crate::representation::types::{
    CrossModelResult, EngineCapabilities, Representation, SimilarityStrategy,
};
use std::future::Future;

/// Representation engine trait.
/// Abstracts embedding generation and comparison.
pub trait RepresentationEngine: Send + Sync {
    /// Model identifier
    fn model_id(&self) -> &str;

    /// Engine capabilities
    fn capabilities(&self) -> &EngineCapabilities;

    /// Encode content to representation
    fn encode(&self, content: &str) -> impl Future<Output = Result<Representation, EngineError>> + Send;

    /// Batch encode
    fn encode_batch(&self, contents: &[&str]) -> impl Future<Output = Result<Vec<Representation>, EngineError>> + Send;

    /// Compute similarity
    fn similarity(&self, a: &Representation, b: &Representation, strategy: SimilarityStrategy) -> f32;

    /// Cross-model comparison
    fn cross_model_similarity(&self, a: &Representation, b: &Representation) -> CrossModelResult;

    /// Serialize representation
    fn serialize(&self, rep: &Representation) -> Vec<u8>;

    /// Deserialize representation
    fn deserialize(&self, bytes: &[u8]) -> Result<Representation, EngineError>;

    /// Check if representation is native to this engine
    fn is_native(&self, rep: &Representation) -> bool;
}

#[derive(Debug, thiserror::Error)]
pub enum EngineError {
    #[error("encoding failed: {0}")]
    EncodingFailed(String),
    #[error("deserialization failed: {0}")]
    DeserializationFailed(String),
    #[error("model not initialized")]
    NotInitialized,
}
